### PR TITLE
Feature/15606 add generic auditing

### PR DESCRIPTION
**Pull request by MichaelRobinson-BJSS**

### CODE REVIEW

File: tests/NHSD.GPIT.BuyingCatalogue.EntityFramework.UnitTests/Identity/IdentityServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            mockAccessor.Setup(a => a.HttpContext).Returns(mockContext.Object);
+
+            var service = new IdentityService(mockAccessor.Object);
+
+            (Guid userId, string userName) = service.GetUserInfo();
+
+            Assert.Equal("Bill Smith", userName);
+            Assert.Equal(testUserId, userId);
+        }
+    }
```
Comment (bjss-chris) [inline tests/NHSD.GPIT.BuyingCatalogue.EntityFramework.UnitTests/Identity/IdentityServiceTests.cs:51]: We can make use of AutoFixture for these tests, like this:

```c#
public static class IdentityServiceTests
{
    [Fact]
    public static void Constructors_VerifyGuardClauses()
    {
        var fixture = new Fixture().Customize(new AutoMoqCustomization());
        var assertion = new GuardClauseAssertion(fixture);
        var constructors = typeof(IdentityService).GetConstructors();

        assertion.Verify(constructors);
    }

    [Theory]
    [CommonAutoData]
    public static void Constructor_NoContext_ThrowsInvalidOperationException(
        [Frozen] Mock<IHttpContextAccessor> mockAccessor,
        IdentityService service)
    {
        mockAccessor.Setup(a => a.HttpContext).Returns((HttpContext)null);

        Assert.Throws<InvalidOperationException>(() => _ = service.GetUserInfo());
    }

    [Theory]
    [CommonAutoData]
    public static void GetUserInfo_GetsValue(
        Mock<HttpContext> mockContext,
        [Frozen] Mock<IHttpContextAccessor> mockAccessor,
        IdentityService service)
    {
        var testUserId = Guid.NewGuid();

        var identity = new ClaimsIdentity(new Claim[]
        {
            new(IdentityService.UserDisplayName, "Bill Smith"),
            new(IdentityService.UserId, testUserId.ToString())
        }, "mock");

        mockContext.Setup(c => c.User).Returns(new ClaimsPrincipal(identity));
        mockAccessor.Setup(a => a.HttpContext).Returns(mockContext.Object);

        (Guid userId, string userName) = service.GetUserInfo();

        userName.Should().Be("Bill Smith");
        userId.Should().Be(testUserId);
    }
}
```

We'll need to add the NuGet package `AutoFixture.Idioms` to the project for the constructor test to work.

**Review by bjss-chris — CHANGES_REQUESTED** [review]

File: src/NHSD.GPIT.BuyingCatalogue.EntityFramework/Identity/IdentityService.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+            var userId = Guid.Parse(GetClaimValue(context.HttpContext.User, UserId));
+            var userName = GetClaimValue(context.HttpContext.User, UserDisplayName);
+
+            return (userId, userName);
+        }
+
+        private static string GetClaimValue(ClaimsPrincipal user, string claimType)
+        {
+            var claim = user.Claims.FirstOrDefault(x => x.Type.Equals(claimType, StringComparison.InvariantCultureIgnoreCase));
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.EntityFramework/Identity/IdentityService.cs:34]: Done

File: src/NHSD.GPIT.BuyingCatalogue.EntityFramework/Identity/IdentityService.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+            return claim != null ? claim.Value : string.Empty;
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.EntityFramework/Identity/IdentityService.cs:36]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.EntityFramework.UnitTests/Identity/IdentityServiceTests.cs (hunk 1 above)
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.EntityFramework.UnitTests/Identity/IdentityServiceTests.cs:51]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.E2E.PublicBrowseTests/Database/IdentityServiceStub.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using NHSD.GPIT.BuyingCatalogue.EntityFramework.Identity;
+
+namespace NHSD.GPIT.BuyingCatalogue.E2ETests.Database
+{
+    public class IdentityServiceStub : IIdentityService
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.E2E.PublicBrowseTests/Database/IdentityServiceStub.cs:6]: Done

File: src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/ClaimsPrincipalExtensions.cs [hunk 5]
```diff
@@ -8,32 +8,32 @@ public static class ClaimsPrincipalExtensions
...
         public static string GetPrimaryOdsCode(this ClaimsPrincipal user)
         {
-            return GetClaimValue(user, "primaryOrganisationOdsCode");
+            return GetClaimValue(user, Constants.Claims.PrimaryOrganisationOdsCode);
         }
 
         public static string[] GetSecondaryOdsCodes(this ClaimsPrincipal user)
         {
-            return user.Claims.Where(x => x.Type.EqualsIgnoreCase("secondaryOrganisationOdsCode")).Select(x => x.Value).ToArray();
+            return user.Claims.Where(x => x.Type.EqualsIgnoreCase(Constants.Claims.SecondaryOrganisationOdsCode)).Select(x => x.Value).ToArray();
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/ClaimsPrincipalExtensions.cs:26]: Done

File: src/NHSD.GPIT.BuyingCatalogue.Framework/Constants/Claims.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿namespace NHSD.GPIT.BuyingCatalogue.Framework.Constants
+{
+    public class Claims
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.Framework/Constants/Claims.cs:3]: Done

**Review by bjss-chris — DISMISSED** [review]

**Pull request merged**